### DISCUSSION TITLE
Moving to libbpf

### DISCUSSION THREAD

**Discussion by yanivagman** [aquasecurity/tracee#260] [Development]
I recently opened a PR making major changes to tracee:
https://github.com/aquasecurity/tracee/pull/250

This PR removes the dependancy of Tracee on gobpf and bcc, and instead moves to using libbpf (which is part of the linux kernel sources).

As such, I want us to merge this change only after we gain some confidence that it works on as many environments as possible.

My request is simple, and I'll appreciate if you can pull the changes from the above branch, and report back if you notice any change - whether it's compilation issues, bpf verifier issues, performance issues, or features not working. If everything works - please report as well.
If you can, please also provide details about the environment you are using - distro name, kernel version and clang version used to compile tracee

Thanks

**Comment by itaysk** [comment]
Updating that we have created a branch called `libbpf` for ongoing work on this. 
Even if you just tell us "it works for me" it will be helpful. 
```bash
git fetch origin
git checkout libbpf
make build
sudo tracee
```

**Reply by grantseltzer** [reply]
I'd add `git submodule update --init` to this list before `make build`

**Comment by grantseltzer** [comment]
Upon running the above commands: 
```
...
clang -S \
	-D __BPF_TRACING__ -D __KERNEL__ -D__TARGET_ARCH_x86 \
	-I dist/libbpf/usr/include \
	-include /lib/modules/5.8.14-300.fc33.x86_64/build/include/linux/kconfig.h \
	-I /lib/modules/5.8.14-300.fc33.x86_64/build/arch/x86/include \
	-I /lib/modules/5.8.14-300.fc33.x86_64/build/arch/x86/include/uapi \
	-I /lib/modules/5.8.14-300.fc33.x86_64/build/arch/x86/include/generated \
	-I /lib/modules/5.8.14-300.fc33.x86_64/build/arch/x86/include/generated/uapi \
	-I /lib/modules/5.8.14-300.fc33.x86_64/build/include \
	-I /lib/modules/5.8.14-300.fc33.x86_64/build/include/uapi \
	-I /lib/modules/5.8.14-300.fc33.x86_64/build/include/generated \
	-I /lib/modules/5.8.14-300.fc33.x86_64/build/include/generated/uapi \
	-I 3rdparty/include  \
	-Wno-address-of-packed-member \
	-Wno-compare-distinct-pointer-types \
	-Wno-deprecated-declarations \
	-Wno-gnu-variable-sized-type-not-at-end \
	-Wno-pointer-sign \
	-Wno-pragma-once-outside-heade \
	-Wno-unknown-warning-option \
	-Wno-unused-value \
	-Wunused \
	-Wall \
	-fno-stack-protector \
	-fno-jump-tables \
	-fno-unwind-tables \
	-fno-asynchronous-unwind-tables \
	-x c \
	-nostdinc \
	-O2 -emit-llvm -c -g tracee/tracee.bpf.c -o dist/tracee.bpf.ll
<built-in>:1:10: fatal error: '/lib/modules/5.8.14-300.fc33.x86_64/build/include/linux/kconfig.h' file not found
#include "/lib/modules/5.8.14-300.fc33.x86_64/build/include/linux/kconfig.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [Makefile:61: dist/tracee.bpf.o] Error 1
```

```
fedo grant in ~> tracee on ~> libbpf
[*] uname -a                                                                                                                                                                                                                              2 ↵
Linux localhost.localdomain 5.8.14-300.fc33.x86_64 #1 SMP Wed Oct 7 21:44:23 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

At least on my system (Fedora 33) `/lib/modules/<kernel>/build` is a link to `/usr/src/kernels/<kernel>`, perhaps there's an issue related to evaluating that?

**Reply by yanivagman** [reply]
Interesting.
Does the path /usr/src/kernels/5.8.14-300.fc33.x86_64 exists?

Do bcc programs work on your environment?
From what I see in bcc sources, they also include kconfig from the same location when compiling the code:
https://github.com/iovisor/bcc/blob/c31e9d6c80f70ff39dff14a7ab3ebec07bec0c7f/src/cc/frontends/clang/kbuild_helper.cc#L119

**Reply by grantseltzer** [reply]
ah missed that:

```
[*] ls -lah /lib/modules/5.8.14-300.fc33.x86_64/build
lrwxrwxrwx. 1 root root 39 Oct  7 17:59 /lib/modules/5.8.14-300.fc33.x86_64/build -> /usr/src/kernels/5.8.14-300.fc33.x86_64
```
5.8.14 doesn't exist, but 5.8.16 does:
```
[*] ls /usr/src/kernels                                                                                         
5.8.16-300.fc33.x86_64
```

Though the bcc tools do work, wonder why the disconnect between /usr/src/kernels and the kernel I have loaded.  Do you know what installs the kernel source to `/usr/src/kernels`/`lib/modules/...`?  Is `/lib/modules/<kernel_version>/build` part of a fs that normally points to the kernel source code of the loaded kernel?

**Reply by yanivagman** [reply]
I think that the kernel headers are installed in /usr/src/kernels as part of the kernel headers packages in your distro, but I'm not sure.
I used to have problems when I updated the kernel and didn't reboot - in these cases the kernel headers were newer than the running kernel and bcc tools didn't work.
First, I would check if the installed kernel matches the kernel headers package.
What is interesting is that you say that bcc tools do work. My guess is that your kernel is configured with CONFIG_IKHEADERS, which was added since 5.2. If that is the case, bcc can find the headers under the sysfs dir /sys/kernel/kheaders.tar.xz - can you please confirm that?

If so, we should do something similar to:
https://github.com/iovisor/bcc/blob/c31e9d6c80f70ff39dff14a7ab3ebec07bec0c7f/src/cc/frontends/clang/kbuild_helper.cc#L200

**Reply by grantseltzer** [reply]
@yanivagman I do not have `/sys/kernel/kheaders.tar.xz`, I can't confirm if my current kernel was compiled with `CONFIG_IKHEADERS` as it wasn't compiled with `CONFIG_IKCONFIG`. Using base fedora 33 install, haven't compiled with any custom options though I was going to soon. 

I have `kernel-headers-5.8.11-300.fc33.x86_64` installed

**Reply by itaysk** [reply]
> I have kernel-headers-5.8.11-300.fc33.x86_64 installed

I believe you are running 5.8.14, as the example you shared:

> fedo grant in ~> tracee on ~> libbpf
> [*] uname -a                                                                                                                                                                                                                              
> Linux localhost.localdomain 5.8.14-300.fc33.x86_64 #1 SMP Wed Oct 7 21:44:23 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

can you verify your kernel headers package is up to date with the kernel version you are running?

**Reply by grantseltzer** [reply]
If I'm not mistaken kernel-headers-5.8.11 is the latest kernel-headers package release for the 5.8 stable kernel. 

Whats the thought process for building using the host computers headers instead of generating a `vmlinux.h`and including that?

**Reply by yanivagman** [reply]
@grantseltzer the kernel headers package version and the running kernel version should match. This is why the build fails.

Regrading your question - at this stage we don't generate vmlinux.h, and require the user to supply the correct kernel headers.
The reason for that, is that generating vmlinux.h requires the kernel to have BTF support (CONFIG_DEBUG_INFO_BTF), or generating it using pahole with the host's vmlinux image. As many kernels not yet configured with BTF at this stage, we don't want to limit the range of supported kernels. We also don't want to request the user to supply the vmlinux image as required when generating vmlinux.h using pahole.

For this reason, we currently move to libbpf, but not yet support CO-RE. This is a mid-term solution until kernels with BTF support will become mainstream.

**Reply by grantseltzer** [reply]
Ah very interesting, that makes sense. I'm upgrading my kernel at the moment and will report back when my headers package matches my running kernel version. Though it may be good to think of a workaround for this as the linux-headers package doesn't get a new version for every new kernel version.

**Reply by yanivagman** [reply]
I still wonder how bcc tools worked on your environment.
If kernel and headers didn't match, and as you say your kernel is not compiled with CONFIG_IKHEADERS - how was bcc able to compile the code? It must have got the correct headers somehow...

**Reply by grantseltzer** [reply]
I believe `CONFIG_IKHEADERS=m` was enabled though the `/sys/kernel/kheaders.tar.xz` did not exist. Whic bcc tools are you talking about compiling specifically?

**Reply by grantseltzer** [reply]
Ok upgraded kernel to 5.9.2 with correct headers and everything built/ran perfectly :)

**Comment by yanivagman** [comment]
Libbpfgo is now part of the master branch.
Any issue related to this, should now be opened as a regular issue
